libraries/AP_InertialSensor/AP_InertialSensor_Invensensev3.cpp

```diff
@@ -540,44 +540,50 @@ void AP_InertialSensor_Invensensev3::read_fifo()
 #else
     const uint8_t fifo_sample_size = INV3_SAMPLE_SIZE;
 #endif
-    if (!block_read(reg_counth, (uint8_t*)&n_samples, 2)) {
-        goto check_registers;
-    }
-
-    if (n_samples == 0) {
-        /* Not enough data in FIFO */
-        goto check_registers;
-    }
 
-    // adjust the periodic callback to be synchronous with the incoming data
-    // this means that we rarely run read_fifo() without updating the sensor data
-    dev->adjust_periodic_callback(periodic_handle, backend_period_us);
-
-    while (n_samples > 0) {
-        uint8_t n = MIN(n_samples, INV3_FIFO_BUFFER_LEN);
-        if (!block_read(reg_data, (uint8_t*)fifo_buffer, n * fifo_sample_size)) {
-            goto check_registers;
+    do {
+        uint8_t n;
+        void *fifo_result = nullptr;
+        if (n_samples == 0) {
+            if (!block_read(reg_counth, (uint8_t*)fifo_buffer, 2 + fifo_sample_size)) {
```

Comment:
should be using _dev->transfer() like in InvensenseV2 and V1, to save the stack space and copy, and preventing a watchdog